main/icicle-repl.hs

```diff
@@ -0,0 +1,113 @@
+{-# LANGUAGE PatternGuards #-}
+{-# LANGUAGE ViewPatterns  #-}
+
+import           Control.Monad
+import           Data.Text                   (Text)
+import qualified Data.Text                   as T
+import qualified System.Console.Haskeline    as HL
+import qualified Text.PrettyPrint.Leijen     as PP
+
+import qualified Icicle.Core.Program.Program as CP
+import qualified Icicle.Repl                 as SR
+import qualified Icicle.Source.Parser        as SP
+
+main :: IO ()
+main = runRepl
+
+runRepl :: IO ()
+runRepl
+  = do putStrLn "welcome to iREPL"
+       HL.runInputT HL.defaultSettings $ loop defaultState
+  where
+    loop :: ReplState -> HL.InputT IO ()
+    loop state
+      = do line <- HL.getInputLine "> "
+           case line of
+             Nothing      -> loop state
+             Just ":quit" -> return ()
+             Just ":q"    -> return ()
+             Just str     -> handleLine state str >>= loop
+
+--------------------------------------------------------------------------------
+
+data ReplState
+   = ReplState
+   { dataFeat :: Maybe String
+   , hasType  :: Bool
+   , hasCore  :: Bool
+   , hasEval  :: Bool }
+
+-- | Settable REPL states
+data Set
+   = ShowType
+   | ShowCore
+   | ShowEval
+
+-- | REPL commands
+data Command
+   = CommandBlank
+   | CommandSet  Set
+   | CommandLoad FilePath
```

Comment:
I don't know whether it makes sense, but I wonder whether it would be useful having any of this stuff in the library proper? would it make it easier to write eg web versions?